contracts/Strategy.sol

```diff
@@ -53,80 +141,243 @@ contract Strategy is BaseStrategy {
+
+        uint256 shares = 0;
+
+        if (wantBalance > dustThreshold) {
+            (, shares) = bentoBox.deposit(
+                BIERC20(address(want)),
+                address(this),
+                address(this),
+                wantBalance,
+                0 // setting this to 0, let's the previous argument determine the deposit size
+            );
+        }
+
+        uint256 sharesInBento = sharesInBento();
+
+        if (sharesInBento > wantToBentoShares(dustThreshold, false)) {
+            bentoBox.transfer(
+                BIERC20(address(want)),
+                address(this),
+                address(kashiPair),
+                sharesInBento
+            );
+
+            kashiPair.addAsset(address(this), true, sharesInBento);
+        }
     }
 
     function liquidatePosition(uint256 _amountNeeded)
         internal
         override
         returns (uint256 _liquidatedAmount, uint256 _loss)
     {
-        // TODO: Do stuff here to free up to `_amountNeeded` from all positions back into `want`
-        // NOTE: Maintain invariant `want.balanceOf(this) >= _liquidatedAmount`
-        // NOTE: Maintain invariant `_liquidatedAmount + _loss <= _amountNeeded`
-
-        uint256 totalAssets = want.balanceOf(address(this));
-        if (_amountNeeded > totalAssets) {
-            _liquidatedAmount = totalAssets;
-            _loss = _amountNeeded.sub(totalAssets);
+        uint256 wantBalance = balanceOfWant();
+        if (_amountNeeded > wantBalance) {
+            uint256 amountToFree = _amountNeeded.sub(wantBalance);
+
+            (, uint256 lastAccrued, ) = kashiPair.accrueInfo();
+            if (block.timestamp > lastAccrued) {
+                // We need to call accrue to accurately calculate totalAssets
+                kashiPair.accrue();
+            }
+
+            uint256 deposited = estimatedTotalAssets().sub(wantBalance);
+
+            if (deposited < amountToFree) {
+                amountToFree = deposited;
+            }
+            if (amountToFree > 0) {
+                uint256 sharesToFree = wantToBentoShares(amountToFree, true);
+                uint256 sharesToFreeFromKashi =
+                    sharesToFree.sub(
+                        bentoBox.balanceOf(
+                            BIERC20(address(want)),
+                            address(this)
+                        )
+                    );
+
+                if (sharesToFreeFromKashi > 0) {
+                    kashiPair.removeAsset(
+                        address(this),
+                        bentoSharesToKashiFraction(sharesToFreeFromKashi, true)
+                    );
+                }
+
+                bentoBox.withdraw(
+                    BIERC20(address(want)),
+                    address(this),
+                    address(this),
+                    0,
+                    sharesInBento()
+                );
+            }
+
+            _liquidatedAmount = balanceOfWant();
```

Comment:
you are not writing any value into `_loss`. Are you sure that `_amountNeeded == _liquidatedAmount `?
I would add a:
```
if (_amountNeeded > _liquidatedAmount) {
  _loss = _amountNeeded - _liquidatedAmount;
}
```